src/executables/csql.c

```diff
@@ -3330,3 +3358,168 @@ get_host_ip (unsigned char *ip_addr)
 
   return 0;
 }
+
+static int
+csql_connect (char *argument, CSQL_ARGUMENT * csql_arg)
+{
+  /*dbname can be stored DB name+ @ + remote hostname + \0 */
+  char dbname[MAXLOGNAME + CUB_MAXHOSTNAMELEN + 2];
+  char username[DB_MAX_USER_LENGTH + 1];
+  const char *delim = " @\n";
+  char *save_ptr_strtok = NULL;
+  char *db_name_ptr = NULL;
+  char *user_name_ptr = NULL;
+  char *host_name_ptr = NULL;
+  char *p = NULL;
+  int error_code;
+  CSQL_ARGUMENT csql_new_arg;
+
+  /*verify @hostname */
+  host_name_ptr = strchr (argument, '@');
+  if (host_name_ptr != NULL)
+    {
+      if (*(host_name_ptr + 1) == '\0')
+	{
+	  host_name_ptr = NULL;
+	  csql_Error_code = CSQL_ERR_CONNECT;
+	  return DO_CMD_FAILURE;
+	}
+    }
+
+  strcpy (dbname, argument);
+  strcpy (username, argument);
+
+  /*find db name following the user name */
+  db_name_ptr = strrchr (dbname, ' ');
+  if (db_name_ptr != NULL)
+    {
+      if (*(db_name_ptr + 1) == '\0')
+	{
+	  db_name_ptr = NULL;
+	  csql_Error_code = CSQL_ERR_CONNECT;
+	  return DO_CMD_FAILURE;
+	}
+      else
+	{
+	  db_name_ptr += 1;
+	}
+    }
+
+  user_name_ptr = strtok_r (username, delim, &save_ptr_strtok);
+  if (user_name_ptr == NULL)
+    {
+      csql_Error_code = CSQL_ERR_CONNECT;
+      return DO_CMD_FAILURE;
+    }
+
+  memset (&csql_new_arg, 0, sizeof (CSQL_ARGUMENT));
+  memcpy (&csql_new_arg, csql_arg, sizeof (CSQL_ARGUMENT));
+
+#if defined(CS_MODE)
+  memset (boot_Host_connected, 0, sizeof (boot_Host_connected));
+#endif /* CS_MODE */
+
+  /*Failed to access other host or db and then access formal db_name */
+  if (db_name_ptr == NULL)
+    {
+      db_name_ptr = strdup (csql_arg->db_name);
```

Comment:
The memory of db_name_ptr needs to be freed.
However, be careful when freeing (see lines 3425 and 3393).
So, it seems necessary to consider whether memory should be allocated to db_name_ptr.

Reply:
I think 
```suggestion
      strcpy (db_name_ptr, csql_arg->db_name);
```
 And then, there will be necessary to user() free.
 @airnet73 How do you think about this?

Reply:
db_name_ptr is of char * type. Also, strcpy cannot be performed because db_name_ptr is null.

Reply:
Right, I will de-allocate this variable. Thank you.